apis/v1alpha1/gatewayclass_types.go

```diff
@@ -64,19 +64,12 @@ type GatewayClassSpec struct {
 	Controller string `json:"controller" protobuf:"bytes,1,opt,name=controller"`
 
 	// AllowedGatewayNamespaceSelector is a selector of namespaces that Gateways
-	// can use this GatewayClass from. This is a standard Kubernetes
-	// LabelSelector, a label query over a set of resources. The result of
-	// matchLabels and matchExpressions are ANDed. Controllers must not support
-	// Gateways in namespaces outside this selector.
+	// of this class can be created in. Gateway validation will prevent Gateways
+	// specifying this class from being created outside of these namespaces.
```

Comment:
Validation webhooks are cool but can we ALWAYS assume that validation webhook will always be setup if these CRDs are in use?

Reply:
I agree that this is not something we can guarantee, but I think it's our least bad option here. Requiring controllers to enforce this gets really messy and leaves open the possibility that a change here could invalidate/break all Gateways using the class. I do think that we'll want to rely on a validating webhook for more than just this use case though, so it would be good for us to make it easy to install and use.

Reply:
Perhaps there's 2 levels of capability here:

1. web hook (may be absent) must enforce the allowed namespaces
2. controllers must ignore gateways that are not in the allowed namespaces

Reply:
I'd originally planned for 2 as well, but as I've thought about it more I've realize it's quite dangerous. That's especially true for any implementations where provisioning Gateways takes time. If this field changed and invalidated a number of Gateways, implementations would have to drop all of them with this interpretation. That would make a change on GatewayClass potentially quite destructive and seems to go against the guiding principle of not breaking things that are working. 

That's where this proposal comes from - move this to a validation problem and something that's only enforced on create. We still gain the security/authorization capabilities here, but we don't need to deal with the potential of a GatewayClass change resulting in removing/destroying a large number of Gateways.

Reply:
I do get the motivation here but this is a divergence from the behavior of this API in other places, which is not necessarily a bad thing.

If we accept this change, my question then is: Why is changing label selector or `onlySameNamespace` not considered to be equally destructive?

Reply:
That's a great point. My main concern is that for some implementations provisioning Gateways is much more expensive than provisioning Routes. As an example, recreating a Gateway may result in a new IP being assigned. I think this specific selector is the only one that affects the validity of Gateways. Everything else affects which Routes a Gateway selects, a resource that seems much simpler for all variations to provision. 

With that said, I do not like the divergence of behaviors here. Ideally both would be treated equally. Unfortunately I can't find a great way to provide the same kind of meaningful create-only validation here for RouteNamespaces. Here's one option that could work, but it relies very heavily on cross-resource validation and would have to deal with quite a few potential race conditions:

On Gateway create:
- If RouteNamespaces are specified, ensure they are at least as restrictive as GatewayClass field.
- If not specified, copy values from GatewayClass field.

On Gateway update:
- Ensure RouteNamespaces are at least as restrictive as GatewayClass field.

On GatewayClass update:
- Ensure RouteNamespaces are at least as restrictive for Gateways of this class. If not, require Gateways to be individually updated first.

To try to match that behavior, we could prevent GatewayClass updates that would change allowedGatewayNamespaces to a value that would invalidate any existing Gateways. Of course changing labels of individual namespaces could still bypass this validation.

All of the above feels relatively fragile, and I'm concerned about the potential race conditions here as well as the significant dependency on webhooks. It could potentially make things feel a bit more consistent here though.

Reply:
Thanks for that write up, that helped me understand the differences in how the gateway and route selector will behave operationally.

So, the problem we are trying to solve here is:

> My main concern is that for some implementations provisioning Gateways is much more expensive than provisioning Routes. As an example, recreating a Gateway may result in a new IP being assigned. I think this specific selector is the only one that affects the validity of Gateways. Everything else affects which Routes a Gateway selects, a resource that seems much simpler for all variations to provision.

which is a problem for all vendors.

Another problem with the change in this PR is that there is no way for a controller to determine if a Gateway that is already created is valid or not:

1. A Gateway passes validation and is running happily, a user updates GatewayClass, making the Gateway invalid.
2. A Gateway is created in an invalid namespace (no webhook or a data-race between gw and gwc create/update operations). The controller starts up and works happily (as per the change in this PR, the controller doesn't need to bother to check if it violates `allowedGatewayNamespaceSelector` in GatewayClass or not. ).

I think this PR is in the right direction and we should explicitly call the following out in docs:
- `allowedGatewayNamespaceSelector` is a field only for validation (this PR)
- If a user is able to bypass validations and all the guardrails and is able to provision a Gateway that is created in namespaces outside `allowedGatewayNamespaceSelector`, the gateway will work.
- In case a Gateway violates `allowedGatewayNamespaceSelector`, the status of the Gateway should reflect so. Meaning, the gateway will continue to serve traffic but it will also have an orphaned or some such condition set to true.

I think that this is similar to Persistent volumes and namespaces. If namespaces in k8s are deleted, PVs that are provisioned in GKE/EKS are not deleted.
In the current case, a namespace deletion will certainly delete the Gateway resource unless we use finalizers.

Reply:
> In case a Gateway violates allowedGatewayNamespaceSelector, the status of the Gateway should reflect so. Meaning, the gateway will continue to serve traffic but it will also have an orphaned or some such condition set to true.

In this case, does the controller still update the violating Gateway? If so. it sounds like we could ship a controller that just monitors the allowed namespaces and updates status on gateways.

Reply:
Those are all great points. I agree that adding a condition to Gateway status would be ideal if it is violating `allowedGatewayNamespaceSelector`. Unfortunately the specifics of that implementation make me hesitant to require this condition to be set. Maybe for the sake of this PR we can provide the condition and say that it _may_ be set on a Gateway if it is violating the `allowedGatewayNamespaceSelector`. 

Requiring implementations to set this condition could be problematic if a controller does not have visibility to all Gateways in all namespaces. On the other hand, developing a common controller in addition to a common webhook would result in significant deployment overhead and additional potential failure modes. I think I'd prefer to revisit requiring this condition to be set after our v1alpha1 release. As part of that discussion we could further explore if a shared implementation makes sense.

Reply:
> Requiring implementations to set this condition could be problematic if a controller does not have visibility to all Gateways in all namespaces. On the other hand, developing a common controller in addition to a common webhook would result in significant deployment overhead and additional potential failure modes.

+1. However, I do feel that sharing code across implementations for these cases will help in further having a better user experience and also lower the development and maintenance cost for all of us.


> Unfortunately the specifics of that implementation make me hesitant to require this condition to be set. Maybe for the sake of this PR we can provide the condition and say that it may be set on a Gateway if it is violating the allowedGatewayNamespaceSelector.

Can you elaborate? I want to understand the concern here.

> Requiring implementations to set this condition could be problematic if a controller does not have visibility to all Gateways in all namespaces.

That shouldn't be the case. If a Gateway resource is created in a cluster, a controller must have access to it, else how would that resource even be satisfied?
If a Gateway is created and no controller is aware of it, isn't that equivalent to that resource not being present in the first place?

Reply:
>If a Gateway is created and no controller is aware of it, isn't that equivalent to that resource not being present in the first place?

Let's say a controller is deployed with access to Gateways in namespaces labeled `env=dev` and it implements a GatewayClass that also limits where Gateways of this class can be created to the same subset of namespaces. What happens if a Gateway referencing that class is created in a namespace outside that subset of namespaces? The controller wouldn't know and therefore wouldn't be able to add `ConditionForbiddenNamespaceForClass` to the status for that Gateway. 

If we require controller implementations to set this condition, it means that controllers would need to have access to Gateways in all namespaces. That leaves a shared/common implementation here which could be helpful but of course requires us to test and maintain one more component as well as requiring users of this API to deploy and maintain one more component. 

All of this for a condition that provides relatively minimal value. I think we're all in agreement on this:
>If a user is able to bypass validations and all the guardrails and is able to provision a Gateway that is created in namespaces outside allowedGatewayNamespaceSelector, the gateway will work.

And if that's the case, this condition would really just indicate that "this Gateway shouldn't really be here but it will keep on working anyway".  Maybe instead we should explore adding a condition on Gateway that indicates that a controller is implementing the Gateway.

Reply:
> Let's say a controller is deployed with access to Gateways in namespaces labeled env=dev and it implements a GatewayClass that also limits where Gateways of this class can be created to the same subset of namespaces. What happens if a Gateway referencing that class is created in a namespace outside that subset of namespaces? The controller wouldn't know and therefore wouldn't be able to add ConditionForbiddenNamespaceForClass to the status for that Gateway.

Correct. But in this case, the Gateway won't even function because the controller can't reach it. The Gateway resource will exist but nothing should happen in the infrastructure.

The point I'm trying to make is, if a controller is controlling/implementing a Gateway and it sees that the Gateway violates `allowedGatewayNamespaceSelector` then the controller MUST add in the ConditionForbiddenNamespaceForClass to the status. This of course doesn't apply to Gateway that is not managed by any controller.

Reply:
Got it, so this condition is meant to be more of a warning to the user? It won't actually have any affect right? It feels like it might get a bit confusing for users if that condition is only set if the controller can see the Gateway. This does seems like a bit of an edge case though.

Reply:
I'm not completely convinced a condition here is necessary. I think it would be impossible to guarantee that each controller set the condition since they may not be able to access Gateways in all namespaces. That inconsistency could make the condition less valuable. With that said, I'm interested in what others think here.

/cc @jpeach @danehans

Reply:
I cant totally see that.
Maybe an event makes sense here? My point is to simply indicate to the user that a Gateway is invalid.

We can track this as a separate issue and this should not be a blocker to get this PR in.